hoomd/Trigger.h

```diff
@@ -87,5 +103,176 @@ class PYBIND11_EXPORT PeriodicTrigger : public Trigger
+        uint64_t m_after;
+    };
+
+/** Not trigger
+ *
+ *  Negates any given trigger.
+*/
+class PYBIND11_EXPORT NotTrigger : public Trigger
+    {
+    public:
+        NotTrigger(std::shared_ptr<Trigger> trigger) :
+            Trigger(), m_trigger(trigger) {}
+
+        bool compute(uint64_t timestep)
+            {
+            return !(m_trigger->compute(timestep));
+            }
+
+        /// Get the trigger thats negated
+        std::shared_ptr<Trigger> getTrigger() {return m_trigger;}
+
+        /// Set the trigger to negate
+        void setTrigger(std::shared_ptr<Trigger> trigger) {m_trigger = trigger;}
+
+    protected:
+        std::shared_ptr<Trigger> m_trigger;
+    };
+
+/** And trigger
+ *
+ *  The logical AND between two triggers.
+*/
+class PYBIND11_EXPORT AndTrigger : public Trigger
+    {
+    public:
+        AndTrigger(std::shared_ptr<Trigger> trigger1,
+                   std::shared_ptr<Trigger> trigger2) :
+            Trigger(), m_trigger1(trigger1), m_trigger2(trigger2) {}
+
+        bool compute(uint64_t timestep)
+            {
+            return m_trigger1->compute(timestep) &&
+                   m_trigger2->compute(timestep);
+            }
+
+        /// Get the first trigger
+        std::shared_ptr<Trigger> getTrigger1() {return m_trigger1;}
+
+        /// Set the second trigger
+        void setTrigger1(std::shared_ptr<Trigger> trigger)
+            {
+            m_trigger1 = trigger;
+            }
+
+        /// Get the second trigger
+        std::shared_ptr<Trigger> getTrigger2() {return m_trigger2;}
+
+        /// Set the second trigger
+        void setTrigger2(std::shared_ptr<Trigger> trigger)
+            {
+            m_trigger2 = trigger;
+            }
+
+    protected:
+        std::shared_ptr<Trigger> m_trigger1;
+        std::shared_ptr<Trigger> m_trigger2;
+    };
+
+/** Or trigger
+ *
+ *  The logical OR between two triggers.
+*/
+class PYBIND11_EXPORT OrTrigger : public Trigger
+    {
+    public:
+        OrTrigger(std::shared_ptr<Trigger> trigger1,
+                   std::shared_ptr<Trigger> trigger2) :
+            Trigger(), m_trigger1(trigger1), m_trigger2(trigger2) {}
+
+        bool compute(uint64_t timestep)
+            {
+            return m_trigger1->compute(timestep) ||
+                   m_trigger2->compute(timestep);
+            }
+
+        /// Get the first trigger
+        std::shared_ptr<Trigger> getTrigger1() {return m_trigger1;}
+
+        /// Set the second trigger
+        void setTrigger1(std::shared_ptr<Trigger> trigger)
+            {
+            m_trigger1 = trigger;
+            }
+
+        /// Get the second trigger
+        std::shared_ptr<Trigger> getTrigger2() {return m_trigger2;}
+
+        /// Set the second trigger
+        void setTrigger2(std::shared_ptr<Trigger> trigger)
+            {
+            m_trigger2 = trigger;
+            }
+
+    protected:
+        std::shared_ptr<Trigger> m_trigger1;
```

Comment:
Are `trigger1` and `trigger2` the most natural names for the child triggers?

Reply:
I also wondered this -- I'd prefer to see a refactor for n-way triggers if it's not too tough. This should be straightforward using `<algorithm>` with `all_of` and `any_of`, assuming that you can make pybind work with variable numbers of arguments.

```
#include <algorithm>    // std::all_of
// ...
std::vector<std::shared_ptr<Trigger> > triggers;
triggers.push_back(some_trigger);
triggers.push_back(another_trigger);
triggers.push_back(third_trigger);
// ...
return std::all_of(triggers.begin(), triggers.end(), [](std::shared_ptr<Trigger> trigger){return trigger();}));

Reply:
I added the refactoring. I like the added flexibility it provides.

Reply:
I would prefer `All` and `Any`, but with the `All` filter, I think this is cleaner.